client/post-editor/editor-gutenberg-opt-in-dialog/index.jsx

```diff
@@ -14,39 +14,66 @@ import Gridicon from 'gridicons';
  * Internal dependencies
  */
 import isGutenbergOptInDialogShowing from 'state/selectors/is-gutenberg-opt-in-dialog-showing';
-import { hideGutenbergOptInDialog } from 'state/ui/gutenberg-opt-in-dialog/actions';
+import {
+	hideGutenbergOptInDialog,
+	optInToGutenberg,
+	optOutOfGutenberg,
+} from 'state/ui/gutenberg-opt-in-dialog/actions';
 import { localize } from 'i18n-calypso';
 import Button from 'components/button';
 import Dialog from 'components/dialog';
+import {
+	withAnalytics,
+	composeAnalytics,
+	recordGoogleEvent,
+	recordTracksEvent,
+	bumpStat,
+} from 'state/analytics/actions';
 
 class EditorGutenbergOptInDialog extends Component {
 	static propTypes = {
 		// connected properties
 		translate: PropTypes.func,
 		gutenbergURL: PropTypes.string,
 		isDialogVisible: PropTypes.bool,
-		hideGutenbergOptInDialog: PropTypes.func,
+		hideDialog: PropTypes.func,
+		optIn: PropTypes.func,
+		optOut: PropTypes.func,
+	};
+
+	optInAndHideDialog = () => {
+		this.props.optIn();
+		this.props.hideDialog();
+	};
+
+	optOutAndHideDialog = () => {
+		this.props.optOut();
+		this.props.hideDialog();
 	};
 
 	render() {
-		const { translate, gutenbergURL, isDialogVisible } = this.props;
+		const { translate, gutenbergURL, isDialogVisible, hideDialog } = this.props;
 		const buttons = [
-			<Button key="gutenberg" href={ gutenbergURL } primary>
+			<Button key="gutenberg" href={ gutenbergURL } onClick={ this.optInAndHideDialog } primary>
```

Comment:
Regardless of this potential race condition, `GUTENBERG_OPT_IN` and `GUTENBERG_OPT_IN_DIALOG_IS_SHOWING` are still dispatched correctly upon entering Gutenberg.

Though, we might improve this by dispatching a `navigate` action in the `onClick` instead of relying on the `href`.

```es6
import { navigate } from 'state/ui/actions';

// ...

optInAndOpenGutenberg = () => {
  this.props.optIn();
  this.props.hideDialog(); // needed because otherwise it will open again on next load 🤔
  this.props.goToGutenberg( this.props.gutenbergURL );
}

// ...

goToGutenberg: gutenbergURL => dispatch( navigate( gutenbergURL ) );
```

Note that we could also streamline the `goToGutenberg` dispatch by providing the URL to it via `mergeProps` (probably overkill tho 🤷‍♂️ )

Reply:
>  It feels like we can dispatch one, and move the side-effect to middleware (either by setting middleware or dispatching another action)

Maybe we can leave this until a later PR that adds a proper `GUTENBERG_OPT_IN` action.